karate-core/src/main/java/com/intuit/karate/core/ScenarioRuntime.java

```diff
@@ -262,12 +262,32 @@ protected void logError(String message) {
             // karate principle: parent variables are always "visible"
             // so we inject the parent variables
             // but they will be over-written by what is local to this scenario
-            if (!caller.isSharedScope()) {
+
+            if (caller.isSharedScope()) {
+                map.putAll(caller.parentRuntime.magicVariables);
```

Comment:
@ptrthomas wasn't sure about impacts for shared scope so decided to keep as-is for now and if I have time I'll think about recreating some additional scenarios

Reply:
@ptrthomas been thinking about this

I think to handle this for calls in shared scope the solution should be to rehydrate in `ScenarioEngine.callFeature()` after the callFeature is done.

I know I’ll also need to add a unit test to show the issue.

Let me know your thoughts.

Reply:
maybe. if you can get a test to fail, sure

Reply:
Just adding the following two lines to `js-called.feature` makes it fail:

```
  * karate.call(true, 'dummy.feature')
  * utils.sayHello()
````

We need to think through some more options as ideally I'd also like to rehydrate **only** the variables that are JS functions - i.e. hopefully there's a way to keep references to what needs to be rehydrated and not rehydrate all variables all the time. Not easy as the references would technically be the keys of a map (not sure if keeping a list of `Map.Entry` of the key/value pairs that have to be rehydrated in the engine would work.. maybe)

Reply:
fix here without further thinking on how to reduce the scope of iterations of the "rehydration" methods
https://github.com/intuit/karate/pull/1756

Reply:
I think we are good for now, hoping we can avoid this when the graal improvements (being able to share JS across contexts) lands